### ISSUE TITLE
Spurious validation claim of VkFence signaled

### ISSUE THREAD

**Issue by buzmeg** [KhronosGroup/Vulkan-ValidationLayers#4712]
**Describe the Issue**
I am irrregularly seeing a VkFence validation error:
```
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0x9fde6b0000000014, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0x9fde6b0000000014[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
```

The problem is that literally 2 lines above my vkQueueSubmit2 calls I have a check assert for that VkFence:
```
            assert(vk.vkGetFenceStatus(uvkd_logical, spfr.uvkf_compute_finished) == vk.VK_NOT_READY);  // Fence unsignaled
            const qq = uvkmvq_swapchain.lock();
            const vkrv = vk.vkQueueSubmit2(qq, avksi2.len, &avksi2, spfr.uvkf_compute_finished);
            uvkmvq_swapchain.unlock();
```
which *succeeds*.  So Vulkan, at least, thinks the fence is going into the vkQueueSubmit2 call unsignalled.

There are only two vkQueueSubmit2 calls in my program, and both are shielded with that assert().  I'm quite mystified as to why that validation layer is firing.

**Valid Usage ID**
Don't expect any warnings to fire.

**Environment:**
 - OS: Ubuntu 22.04.1 LTS
 - GPU: Mesa Intel® UHD Graphics 620 (WHL GT2)
 - SDK or header version if building from repo: Vulkan Instance Version: 1.3.216
 - Options enabled (synchronization, best practices, etc.):  Standard Preset

**Additional context**
Log of program run: (RPC increment of 500 is approximately 15 seconds)

info: RPC count check: rpc_0 8000
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xb991fa0000000024, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xb991fa0000000024[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xb991fa0000000024, type: 7, name: NULL
info: RPC count check: rpc_0 8500
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xd10d270000000018, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xd10d270000000018[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xd10d270000000018, type: 7, name: NULL
info: RPC count check: rpc_0 9000
info: RPC count check: rpc_0 9500
info: RPC count check: rpc_0 10000
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0x2cfba2000000001c, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0x2cfba2000000001c[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0x2cfba2000000001c, type: 7, name: NULL
info: RPC count check: rpc_0 10500
info: RPC count check: rpc_0 11000
info: RPC count check: rpc_0 11500
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xd5b26f0000000010, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xd5b26f0000000010[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xd5b26f0000000010, type: 7, name: NULL
info: RPC count check: rpc_0 12000
info: RPC count check: rpc_0 12500
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0x9fde6b0000000014, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0x9fde6b0000000014[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0x9fde6b0000000014, type: 7, name: NULL
info: RPC count check: rpc_0 13000
info: RPC count check: rpc_0 13500
info: RPC count check: rpc_0 14000
info: RPC count check: rpc_0 14500
info: RPC count check: rpc_0 15000
info: RPC count check: rpc_0 15500
info: RPC count check: rpc_0 16000
info: RPC count check: rpc_0 16500
info: RPC count check: rpc_0 17000
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xab64de0000000020, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xab64de0000000020[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xab64de0000000020, type: 7, name: NULL
info: RPC count check: rpc_0 17500
info: RPC count check: rpc_0 18000
info: RPC count check: rpc_0 18500
info: RPC count check: rpc_0 19000
info: RPC count check: rpc_0 19500
info: RPC count check: rpc_0 20000
info: RPC count check: rpc_0 20500
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xe88693000000000c[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xe88693000000000c, type: 7, name: NULL
info: RPC count check: rpc_0 21000
info: RPC count check: rpc_0 21500
info: RPC count check: rpc_0 22000
info: RPC count check: rpc_0 22500
^Ninfo: RPC count check: rpc_0 23000
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xe88693000000000c[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xe88693000000000c, type: 7, name: NULL
info: RPC count check: rpc_0 23500
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0xb991fa0000000024, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0xb991fa0000000024[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0xb991fa0000000024, type: 7, name: NULL
info: RPC count check: rpc_0 24000
VUID-vkQueueSubmit2-fence-04894(ERROR / SPEC): msgNum: -1295338315 - Validation Error: [ VUID-vkQueueSubmit2-fence-04894 ] Object 0: handle = 0x2cfba2000000001c, type = VK_OBJECT_TYPE_FENCE; | MessageID = 0xb2cab4b5 | vkQueueSubmit2(): VkFence 0x2cfba2000000001c[] submitted in SIGNALED state.  Fences must be reset before being submitted The Vulkan spec states: If fence is not VK_NULL_HANDLE, fence must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.216.0/linux/1.3-extensions/vkspec.html#VUID-vkQueueSubmit2-fence-04894)
    Objects: 1
        [0] 0x2cfba2000000001c, type: 7, name: NULL
info: RPC count check: rpc_0 24500

**Comment by jeremyg-lunarg** [comment]
This seems like it may be related to https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/4271, is it possibly you to try a newer sdk build and see if that fixes the problem?

If that doesn't do it, what does your code to call vkWaitForFences() and vkResetFences() look like?  My first attempt to make a test for this error is so far not reproducing it when run against the 1.3.216 validation layer.

**Comment by buzmeg** [comment]
Erm, why would vkWaitForFences of vkResetFences have any bearing given that I actually check the fence is unsignaled with vkGetFenceStatus two lines before I submit to the queue?

**Comment by buzmeg** [comment]
Okay, I will take a look and see what is required to override the Mesa version of the SDK.  That will not be quick task, though.

Here's the code you requested.  Normally, the VkFence has signalled quite a while prior since I have at least 3 frames worth of resources in flight:

Swapchain thread:
```
                const uvk_frame_resources = auvk_per_frame_resources[u32_render_frame_index];

                if (!ff_refresh_stalled) {
                    //linfo("Frame Index: {}", .{u32_render_frame_index});

                    // This fence should have fired several frames ago--the only case where that might not be true
                    // is when number of frames == number of images.
                    var avkf: [1]vk.VkFence = undefined;
                    avkf[0] = uvk_frame_resources.uvkf_frame_resources_released;

                    var vkrv = vk.vkWaitForFences(uvkd_logical, avkf.len, &avkf, vk.VK_TRUE, 0);
                    if (vkrv == vk.VK_TIMEOUT) {
                        lerr("Resource fence timeout.  Should never fire.", .{});
                    }
                    assert(vkrv == vk.VK_SUCCESS);

                    vkrv = vk.vkResetFences(uvkd_logical, avkf.len, &avkf);
                    assert(vkrv == vk.VK_SUCCESS);
                } else {
                    // Make sure fence didn't escape reset somehow
                    assert(vk.vkGetFenceStatus(uvkd_logical, uvk_frame_resources.uvkf_frame_resources_released) == vk.VK_NOT_READY);  // Fence unsignaled
                }

```

Render thread:
```
        const spfr = auvk_slaggy_per_frame_resources[u64_slaggy_next_frame_index];
        {
            const uvkcb = spfr.auvkcb[0];

            {  // FIXME: Fence wait should be redundant ...
                const avkf = [_]vk.VkFence {
                    spfr.uvkf_compute_finished,
                };

                const VK_FF_WAIT_ALL = vk.VK_TRUE;
                const UVK_UI_TIMEOUT_NS = 1000000;
                const vkrv = vk.vkWaitForFences(uvkd_logical, avkf.len, &avkf, VK_FF_WAIT_ALL, UVK_UI_TIMEOUT_NS);
                if (vkrv != vk.VK_SUCCESS) {
                    lerr("Fence timeout failure on what should be redundant wait", .{});
                }
                assert(vkrv == vk.VK_SUCCESS);
            }



            {
                const avkf = [_] vk.VkFence {
                    spfr.uvkf_compute_finished,
                };
                const vkrv = vk.vkResetFences(uvkd_logical, avkf.len, &avkf);
                assert(vkrv == vk.VK_SUCCESS);
            }

            {
                const vkrv = vk.vkResetCommandPool(uvkd_logical, spfr.uvkcp, 0);
                assert(vkrv == vk.VK_SUCCESS);
            }


```

**Comment by jeremyg-lunarg** [comment]
Thanks,  I'll try to make a unit test based on this info.

**Comment by jeremyg-lunarg** [comment]
Are you still seeing this with the 1.3.236 SDK?  Fixes for a [very similar issue](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/4733) made it into that sdk.

**Comment by spencer-lunarg** [comment]
@buzmeg if you are still seeing this please feel free to reopen, otherwise going to close as this should be fixed with the 1.3.236 SDK now

**Comment by buzmeg** [comment]
I can't really test this easily right now.  If it pops up again, I'll file another report.

Thanks.
